File: ts_distribution/public/js/sales_order.js
// Copyright (c) 2021, Tech Station and contributors
// For license information, please see license.txt

cur_frm.add_fetch('payment_method',  'create_payment_entry_on_so',  'create_payment_entry_on_so');
cur_frm.add_fetch('payment_method',  'create_payment_entry_on_delivery',  'create_payment_entry_on_delivery');
cur_frm.add_fetch('item_code',  'shipping_discount',  'shipping_discount');

frappe.ui.form.on('Sales Order', {
	refresh: function(frm) {
		frm.set_query("payment_method", "multiple_payment", function (doc, cdt, cdn) {
			let d = locals[cdt][cdn];
			return {
				filters: [
					['Mode of Payment', 'name', 'not in', "Paid,Gift"],
				]
			};
		});

		frm.set_query("account", "multiple_payment", function (doc, cdt, cdn) {
			let d = locals[cdt][cdn];
			return {
				query: "ts_distribution.utils.account.getAccount",
				filters: {
					'name': d.payment_method,
					'company': frm.doc.company
				}
			};
		});

		frm.set_query("customer_driver", function () {
			return {
				"filters": {
					"customer": frm.doc.customer,
				}
			};
		});

		update_multipayment_status(frm);
	},
	distribution_type: function(frm) {
		frm.set_value("transportation_vehicle","");
		frm.set_value("vehicle_type","");
		frappe.db.get_value(frm.doc.distribution_type, frm.doc.sales_type, "warehouse", (r) => {
				frm.set_value("set_warehouse",r.warehouse);
		});
	},
	transportation_vehicle: function(frm) {
		frm.set_value("vehicle_type","");
		set_total_discount(frm);
	},
	customer: function(frm) {
		if (frm.doc.customer && frm.doc.company) {
			frappe.call({
				method: "ts_distribution.hook.sales_order.getCredit_Limit",
				args: {
					customer: frm.doc.customer,
					company: frm.doc.company
				},
				callback: function (r) {
					if(r.message){
						frm.set_value("customer_credit_limit", r.message);
					}
					else{
						frm.set_value("customer_credit_limit", 0);
					}
				}
			});
		}
	},
	fetch_advance_payment: function (frm) {
		set_advance_entries(frm);
	},
	before_save: function (frm) {
		if(frm.doc.docstatus === 0){
			set_totals(frm);
		}
	},
});

frappe.ui.form.on("Sales Order Item", {
	"rate": function (frm, cdt, cdn) {
		set_discount(frm, cdt, cdn);
		set_total_discount(frm);
	},
	"qty": function (frm, cdt, cdn) {
		set_discount(frm, cdt, cdn);
		set_total_discount(frm);
	},
	"shipping_discount": function (frm, cdt, cdn) {
		set_discount(frm, cdt, cdn);
		set_total_discount(frm);
	},
	"total_shipping_discount": function (frm, cdt, cdn) {
		set_total_discount(frm);
	}
});

frappe.ui.form.on("TD Multiple Payment", {
	"payment_method": function (frm, cdt, cdn) {
		var d = locals[cdt][cdn];
		if (d.payment_method) {
			frappe.call({
				"method": "ts_distribution.utils.account.getDefaultAccount",
				args: {
					method: d.payment_method,
					company: frm.doc.company
				},
				callback: function (r) {
					frappe.model.set_value(d.doctype, d.name, "account", r.message);
				}
			});

		}
	},
	"amount": function (frm, cdt, cdn) {
		var d = locals[cdt][cdn];
		var paid_amount = 0;
		var unpaid_amount = 0;
		var multiple_payment = frm.doc.multiple_payment;

		for (var i in multiple_payment) {
			if(multiple_payment[i].create_payment_entry_on_so  == 1){
				paid_amount = paid_amount + multiple_payment[i].amount;
			}
			if(multiple_payment[i].create_payment_entry_on_so === 0){
				unpaid_amount = unpaid_amount + multiple_payment[i].amount;
			}	
		}
			frm.set_value("paid_amount", paid_amount);
			frm.set_value("unpaid_amount", unpaid_amount);
			set_totals(frm);
			frm.refresh_fields();

		if((paid_amount + unpaid_amount + frm.doc.advance_amount_adjusted) > frm.doc.rounded_total){
			frappe.model.set_value(d.doctype, d.name, "amount", 0.0);
			frappe.throw(__("Payment amount can not be higher then rounded total <b>"+frm.doc.rounded_total+"</b>"));
		}
	},
	"multiple_payment_remove": function (frm, cdt, cdn) {
		var paid_amount = 0;
		var unpaid_amount = 0;
		var multiple_payment = frm.doc.multiple_payment;

		for (var i in multiple_payment) {
			if(multiple_payment[i].create_payment_entry_on_so  == 1){
				paid_amount = paid_amount + multiple_payment[i].amount;
			}
			if(multiple_payment[i].create_payment_entry_on_so === 0){
				unpaid_amount = unpaid_amount + multiple_payment[i].amount;
			}
		}
			frm.set_value("paid_amount", paid_amount);
			frm.set_value("unpaid_amount", unpaid_amount);
			set_totals(frm);
			frm.refresh_fields();
	}
});

frappe.ui.form.on("TD Advance Customer Payment", {
	"adjust_amount": function (frm, cdt, cdn) {
		var d = locals[cdt][cdn];
		var total = 0;
		var advance = frm.doc.advance_customer_payment;

		if(d.adjust_amount > d.unallocated_amount){
			frappe.model.set_value(d.doctype, d.name, "adjust_amount", 0.0);
			frappe.throw(__("Adjust amount can not be higher then advance amount <b>"+d.unallocated_amount+"</b>"));
		}
		else{
			for (var i in advance) {
				total = total + advance[i].adjust_amount;
			}
				frm.set_value("advance_amount_adjusted", total);
				set_totals(frm);
				frm.refresh_fields();
		}
		if(total > frm.doc.rounded_total){
			frappe.model.set_value(d.doctype, d.name, "adjust_amount", 0.0);
			frappe.throw(__("Total advance adjust amount can not be higher then rounded total <b>"+frm.doc.rounded_total+"</b>"));
		}
	},
	"advance_customer_payment_remove": function (frm, cdt, cdn) {
		var total = 0;
		var advance = frm.doc.advance_customer_payment;

		for (var i in advance) {
			total = total + advance[i].adjust_amount;
		}
			frm.set_value("advance_amount_adjusted", total);
			frm.refresh_fields();
	}
});

function set_advance_entries(frm) {
	if(frm.doc.customer){
		frappe.call({
			"method": "ts_distribution.hook.sales_order.return_unallocated_amount",
			args: {
				customer: frm.doc.customer
			},
			callback: function (r) {
				if (r.message) {
					frm.doc.advance_customer_payment = [];
					for (var payment_entry in r.message) {
						var payment_entry = r.message[payment_entry]
						var a = frm.add_child("advance_customer_payment");
						a.payment_entry = payment_entry.name;
						a.payment_received_date = payment_entry.posting_date;
						a.payment_method = payment_entry.mode_of_payment;
						a.unallocated_amount = payment_entry.unallocated_amount;
						a.paid_to = payment_entry.paid_to;
					}
					frm.refresh_fields();
				}
			}
		});
	}
}

function set_discount(frm, cdt, cdn) {
	var row = locals[cdt][cdn];
	if(row.rate && row.rate < row.shipping_discount){
		frappe.throw(__("Rate Can Not Be Lesser Then Shipping Discount <b>{0}.</b>",[row.shipping_discount]));
	}
	else{
		frappe.model.set_value(cdt, cdn, "total_shipping_discount", row.qty * row.shipping_discount);
	}
}

function set_total_discount(frm) {
	var total_discount = 0;
	var items = frm.doc.items;
	for (var i = 0; i < items.length; i++) {
		total_discount += items[i].total_shipping_discount;
	}
	if(frm.doc.distribution_type == "Indirect Sales" && frm.doc.transportation_vehicle == "Customer Vehicles"){
		frm.set_value("discount_amount", total_discount);
		frm.refresh_fields();
	}
	else{
		frm.set_value("discount_amount", 0.0);
		frm.refresh_fields();
	}
}

function set_totals(frm) {
	frm.set_value("outstanding_amount",frm.doc.rounded_total - (frm.doc.advance_amount_adjusted + frm.doc.paid_amount));
}

function update_multipayment_status(frm) {
	$.each(frm.doc.multiple_payment,  function(i,  d) {
		if(!frm.is_new() && d.payment_method){
		frappe.call({
			method: "ts_distribution.hook.sales_order.get_PaymentEntry_data",
			args: {
				order_no: frm.doc.name,
				mode_of_payment: d.payment_method,
				paid_amount: d.amount,
			},
			callback: function (r) {
				if (r.message == 1 && d.create_payment_entry_on_delivery === 0){
					d.status = "<span class=\"indicator-pill whitespace-nowrap green\"> <span>Paid</span></span>";
				}
				else{
					d.status = "<span class=\"indicator-pill whitespace-nowrap orange\"> <span>Unpaid</span></span>";
				}
				frm.refresh_fields();
			}
		});
		}
		if(d.create_payment_entry_on_delivery === 0 && d.create_payment_entry_on_so == 0){
			d.payment_stage = "<span class=\"indicator-pill whitespace-nowrap red\"><span>Payment Not Specified</span></span>";
		}
		if(d.create_payment_entry_on_delivery === 0 && d.create_payment_entry_on_so == 1){
			d.payment_stage = "<span class=\"indicator-pill whitespace-nowrap green\"><span>Advance Payment</span></span>";
		}
		if(d.create_payment_entry_on_delivery === 1 && d.create_payment_entry_on_so == 0){
			d.payment_stage = "<span class=\"indicator-pill whitespace-nowrap orange\"><span>Payment With Delivery</span></span>";
		}
	});
}

File: ts_distribution/hook/sales_order.py
from __future__ import unicode_literals
import frappe
from frappe import msgprint, throw, _
from frappe.utils import flt, get_url, money_in_words
from frappe.model.document import Document
from erpnext.accounts.utils import reconcile_against_document
from erpnext.accounts.party import get_party_account

def validate(doc,method):
	validate_discount(doc)
	if doc.outstanding_amount > 0:
		if len(doc.multiple_payment) == 0:
			frappe.throw("Please add payment method and account in <b>Multiple Payment</b> table")

def on_submit(doc, method):
	publish_message(doc)
	validate_discount(doc)
	if len(doc.multiple_payment) > 0:
		create_multi_payment_entry(doc)

	if len(doc.advance_customer_payment) > 0:
		updatePaymentEntry(doc)

def validate_discount(doc):
	if doc.distribution_type == "Indirect Sales" and doc.transportation_vehicle == "Customer Vehicles":
		for d in doc.items:
			if d.rate < d.shipping_discount:
				frappe.throw(_("Row <b>{0}</b>, Item Code <b>{1}</b><br><br>Rate Can Not Be Lesser Then \
				Shipping Discount <b>{2}</b>").format(d.idx,d.item_code,d.shipping_discount))

@frappe.whitelist()
def getCredit_Limit(customer,company):
	credit_limit = frappe.db.sql("""select sum(credit_limit) from `tabCustomer Credit Limit` 
    where parent = %s and company = %s;""",(customer,company)) or 0.0
	return credit_limit[0][0] if credit_limit else 0.0

@frappe.whitelist()
def return_unallocated_amount(customer):
	data = []
	payment_entry_list = frappe.db.sql("""select name,posting_date,mode_of_payment,unallocated_amount,paid_to
	from `tabPayment Entry` where unallocated_amount > 0 and docstatus = 1 and 
	party = %s order by posting_date desc;""",(customer),as_dict = True)

	for payment_entry in payment_entry_list:
		return_data = {}
		return_data["name"] = payment_entry.name
		return_data["posting_date"] = payment_entry.posting_date
		return_data["mode_of_payment"] = payment_entry.mode_of_payment
		return_data["unallocated_amount"] = payment_entry.unallocated_amount
		return_data["paid_to"] = payment_entry.paid_to
		data.append(return_data)
	return data

@frappe.whitelist()
def get_PaymentEntry_data(order_no,mode_of_payment,paid_amount):
	return frappe.db.get_value('Payment Entry', {'reference_no': order_no,'mode_of_payment': mode_of_payment,'paid_amount': paid_amount}, ['docstatus'])

def updatePaymentEntry(self):
	lst = []
	for d in self.get('advance_customer_payment'):
		if flt(d.adjust_amount) > 0:
			args = frappe._dict({
				'voucher_type': "Payment Entry",
				'voucher_no': d.payment_entry,
				'voucher_detail_no': d.reference_row,
				'against_voucher_type': self.doctype,
				'against_voucher': self.name,
				'account': get_party_account("Customer",self.customer,self.company),
				'party_type': "Customer",
				'party': self.customer,
				'is_advance': 'Yes',
				'dr_or_cr': "credit_in_account_currency",
				'unadjusted_amount': flt(d.unallocated_amount),
				'allocated_amount': flt(d.adjust_amount),
				'precision': d.precision('adjust_amount'),
				'grand_total': self.rounded_total,
				'outstanding_amount': flt(self.rounded_total - self.advance_paid),
				'difference_account': frappe.db.get_value('Company', self.company, 'exchange_gain_loss_account')
			})
			lst.append(args)

	if lst:
		from erpnext.accounts.utils import reconcile_against_document
		reconcile_against_document(lst)

def create_multi_payment_entry(doc):
	payment_entry_count =  0
	for mpe in doc.multiple_payment:
		if (mpe.amount > 0 and 
			mpe.payment_method and 
			mpe.create_payment_entry_on_so == 1):
			pe = frappe.get_doc({
			"doctype": "Payment Entry",
			"payment_type": "Receive",
			"company": doc.company,
			"cost_center": frappe.db.get_value("Company", doc.company, ["cost_center"]),
			"posting_date": doc.transaction_date,
			"mode_of_payment": mpe.payment_method,
			"party_type": "Customer",
			"party": doc.customer,
			"party_name": doc.customer_name,
			"paid_from": frappe.db.get_value("Company", doc.company, ["default_receivable_account"]),
			"paid_to": mpe.account,
			"paid_amount": mpe.amount,
			"base_paid_amount": mpe.amount,
			"received_amount": mpe.amount,
			"received_amount": mpe.amount,
			"base_received_amount": mpe.amount,
			"reference_no": doc.name,
			"reference_date": doc.transaction_date,
			"delivery_payment_transaction_id": mpe.name,
			"references": [{
				"reference_doctype": "Sales Order",
				"reference_name": doc.name,
				"due_date": doc.transaction_date,
				"total_amount": doc.rounded_total,
				"outstanding_amount": doc.rounded_total,
				"allocated_amount": mpe.amount,
			}],
			})
			pe.insert(ignore_permissions=True,ignore_if_duplicate=True,ignore_links=True,ignore_mandatory=True)
			pe.save()
			pe.submit()
			payment_entry_count += 1
	url = get_url("/app/payment-entry")
	frappe.msgprint(_("{0} <b><a href={1}>Payment Entry</a></b> Created").format(payment_entry_count,url))

def publish_message(doc):
	for d in doc.multiple_payment:
		if d.create_payment_entry_on_so != frappe.db.get_value("Mode of Payment",d.payment_method,"create_payment_entry_on_so"):
			frappe.msgprint(_('Mode of Payment setting changes detected in <b>Row : {0}</b> for Mode of Payment <b>{1}</b>  \
			<br><br>Old changes will not be overrite with new changes.').format(d.idx,d.payment_method))	